File: in_built_data_structures_and_algorithims/sim.py
"""https://ualberta.kattis.com/sessions/jot9wq/problems/sim
"""

import sys
from typing import List


def sim(input: str) -> None:
    result = []
    cursor = 0  # The last vacant position, i.e. the position in which the next char would go
    for char in input:
        if char == '<':
            if cursor > 0:
                result.pop(cursor-1)
                cursor -= 1
        elif char == ']':
            cursor = len(result)
        elif char == '[':
            cursor = 0
        else:
            result.insert(cursor, char)
            cursor += 1
    print(''.join(result))

def main():
    # input_lines = sys.stdin.read().splitlines()
    # num_of_cases = input_lines[0]
    # for input in input_lines[1:]:
    #     sim(input)
    sim('my ]]name]] is]] steva<en]<n halim]]]]]')
    sim('<<hate<<<<loves[steva<en ] cs2040c< and also cs2040c')

if __name__ == '__main__':
    main()

File: in_built_data_structures_and_algorithims/sim.cpp
/* https://ualberta.kattis.com/sessions/jot9wq/problems/sim
*/

#include <iostream>
#include <list>
#include <string>
using namespace std;

int main() {
    int num_of_test_cases;
    cin >> num_of_test_cases;
    cin.ignore();

    while (num_of_test_cases--) {
        string input;
        list<char> result;
        getline(std::cin, input);

        list<char>::iterator iter = result.begin();
        for (char ch : input) {
            if (ch == '<') {
                if (iter != result.begin())
                    iter = result.erase(--iter);
            } else if (ch =='[') {
                iter = result.begin();
            } else if (ch == ']') {
                 iter = result.end();
            } else {
                iter = result.insert(iter, ch);
                iter++;
            }
        }
        for (char ch : result)
            cout << ch;
        cout << endl;
    }
}
